parlai/agents/transformer/modules/interfaces.py

```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/env python3
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+from __future__ import annotations
+from dataclasses import dataclass
+from torch import nn
+from typing import Generic, Optional, Type, TypeVar
+
+
+class ModularComponent(nn.Module):
+    """
+    A transformer component, optionally with swappable subcomponents.
+
+    For an example of how to use this, see parlai/agents/examples/transformer_variant.py
+    """
+
+    @dataclass
+    class Template:
+        """
+        Define any swappable components by adding their class as a parameter of this
+        object.
+        """
+
+
+MC = TypeVar('MC', bound=ModularComponent)
+
+
+# ModularComponentBuilder should be immutable, since it is used in default
+# function parameter values. In python, mutating a param default value changes
+# it for all subsequent invocations.
+class ModularComponentBuilder(Generic[MC]):
+    """
+    When a component has swappable subcomponents, use this object to specify both the
+    component type and it's subcomponent types at the same time.
+    """
+
+    def __init__(
+        self, klass: Type[MC], template: Optional[ModularComponent.Template] = None
+    ) -> None:
+        self._klass = klass
```

Comment:
what is 'klass' (sorry)?

Reply:
It's type-annotated in `__init__` for reference, but it's a class. Spelled with a `k` so it doesn't clash with the python reserved word `class`. So `TransformerEncoderLayer`, for example.